### PR TITLE
Update TeamCity diff checker tool

### PR DESCRIPTION
**Pull request by SarahFrench**
# Description

This PR:
- Adds testing to the TC diff checker
- Refactors logic to add more user feedback and provide feedback about mismatches in both directions between TC config and provider code
- Changes how the tool is invoked from GHAs

## Context

 TeamCity diff checker has been failing for a few weeks:

Screenshot 2024-11-25 at 15 50 55

[Here is an example](https://github.com/GoogleCloudPlatform/magic-modules/actions/runs/12003012549/job/33455826322):
Screenshot 2024-11-25 at 15 51 54

This error is confusing because:

* services_ga.kt only contains [recaptchaenterprise](https://github.com/GoogleCloudPlatform/magic-modules/blob/7e4d09b9905867accec0368e58311b358d4715ca/mmv1/third_party/terraform/.teamcity/components/inputs/services_ga.kt#L604C6-L604C25), not recaptcha
* The GA provider contains only a [services/recaptchaenterprise](https://github.com/hashicorp/terraform-provider-google/tree/main/google/services/recaptchaenterprise) package, and there isn't a [services/recaptcha](https://github.com/hashicorp/terraform-provider-google/tree/main/google/services/recaptcha).

However:
* There is a [services/recaptcha in the Beta provider](https://github.com/hashicorp/terraform-provider-google-beta/tree/main/google-beta/services/recaptcha)
* This is missing in the [services_beta.kt file
](https://github.com/GoogleCloudPlatform/magic-modules/blob/7e4d09b9905867accec0368e58311b358d4715ca/mmv1/third_party/terraform/.teamcity/components/inputs/services_beta.kt#L609)

With better user feedback this could be less confusing

```release-note:none

```

### CODE REVIEW

File: .github/workflows/teamcity-services-diff-check-weekly.yml [hunk 1]
```diff
@@ -37,11 +37,13 @@ jobs:
           run: echo "GOOGLE_BETA_REPO_PATH=${{ env.OUTPUT_PATH}}" >> $GITHUB_ENV
         - name: Check that new services have been added to the TeamCity configuration code
           run: |
-              # Create lists of service packages in providers
-              ls ${{env.GOOGLE_REPO_PATH}}/google/services > tools/teamcity-diff-check/services_ga.txt
-              ls ${{env.GOOGLE_BETA_REPO_PATH}}/google-beta/services > tools/teamcity-diff-check/services_beta.txt
-    
+              # Create lists of service packages in providers. Need to cd into repos where go.mod is to do this command.
+              cd ${{env.GOOGLE_REPO_PATH}}
+              go list -f '{{.Name}}' ${{env.GOOGLE_REPO_PATH}}/google/services/... > $GITHUB_WORKSPACE/provider_services_ga.txt
```
Comment (SarahFrench) [inline .github/workflows/teamcity-services-diff-check-weekly.yml:42]: Using go list allows us to list the actual packages present in the services directory. This protects us a bit against people who name the folders differently from the package name (matching names is what is usual).

**Review by BBBmau — APPROVED** [review]
looks good! Ran the tests locally:
```hcl
└─(10:05:08 on update-teamcity-diff-checker)──> go test -timeout 30s -run ^Test_ -v                                                                                                                                                                   ──(Wed,Nov27)─┘
=== RUN   Test_main_happyPaths
=== RUN   Test_main_happyPaths/something_missing_in_TeamCity_config_present_in_provider_code
=== RUN   Test_main_happyPaths/something_missing_in_provider_code_present_in_TeamCity_config
=== RUN   Test_main_happyPaths/everything_matches
--- PASS: Test_main_happyPaths (0.00s)
    --- PASS: Test_main_happyPaths/something_missing_in_TeamCity_config_present_in_provider_code (0.00s)
    --- PASS: Test_main_happyPaths/something_missing_in_provider_code_present_in_TeamCity_config (0.00s)
    --- PASS: Test_main_happyPaths/everything_matches (0.00s)
=== RUN   Test_main_unhappyPaths
=== RUN   Test_main_unhappyPaths/cannot_find_provider_service_file
=== RUN   Test_main_unhappyPaths/cannot_find_TeamCity_service_file
=== RUN   Test_main_unhappyPaths/empty_TeamCity_service_file
=== RUN   Test_main_unhappyPaths/empty_provider_service_file
--- PASS: Test_main_unhappyPaths (0.00s)
    --- PASS: Test_main_unhappyPaths/cannot_find_provider_service_file (0.00s)
    --- PASS: Test_main_unhappyPaths/cannot_find_TeamCity_service_file (0.00s)
    --- PASS: Test_main_unhappyPaths/empty_TeamCity_service_file (0.00s)
    --- PASS: Test_main_unhappyPaths/empty_provider_service_file (0.00s)
=== RUN   Test_listDifference
=== RUN   Test_listDifference/detects_when_lists_match
=== RUN   Test_listDifference/detects_when_items_from_list_A_is_missing_items_present_in_list_B_-_1_missing
=== RUN   Test_listDifference/detects_when_items_from_list_A_is_missing_items_present_in_list_B_-_2_missing
=== RUN   Test_listDifference/doesn't_detect_differences_if_list_A_is_a_superset_of_list_B
--- PASS: Test_listDifference (0.00s)
    --- PASS: Test_listDifference/detects_when_lists_match (0.00s)
    --- PASS: Test_listDifference/detects_when_items_from_list_A_is_missing_items_present_in_list_B_-_1_missing (0.00s)
    --- PASS: Test_listDifference/detects_when_items_from_list_A_is_missing_items_present_in_list_B_-_2_missing (0.00s)
    --- PASS: Test_listDifference/doesn't_detect_differences_if_list_A_is_a_superset_of_list_B (0.00s)
PASS
ok      github.com/GoogleCloudPlatform/magic-modules/tools/teamcity-diff-check  0.232s
```

**Comment by SarahFrench** [comment]
Thanks! I've manually tested the GHA yaml in my fork, but I'll merge this and then run the weekly check GHA on this repo to confirm all's ok

**Comment by SarahFrench** [comment]
Looks ok: https://github.com/GoogleCloudPlatform/magic-modules/actions/runs/12059681543/job/33628745318

Screenshot 2024-11-27 at 23 27 58

**Pull request merged**